update main + scripts/ README's (#1611)
sparkletown/sparkle: README.md

# Sparkle Web App

Codebase for the Sparkle platform, brought to you by [Sparkle](https://sparklespaces.com/), a platform for the most immersive & interactive virtual events.

### Getting started

What you will need:

- A free Firebase account
- Your favorite command-line interface (e.g. Terminal or zsh on a Mac)

> - Optional: accounts with Stripe, Twillio

Reviewer: fixed here: https://github.com/sparkletown/sparkle/pull/1611/commits/c129ebfd44e8b6bb265ae499b7e81964f93b0fcc

- A little fairy dust

---
<!-- section 1 -->

## Part One: Firebase Project Setup

> ### Step 1: Create New Firebase Project

Reviewer 2: @whoaitsaimz This is WAY too detailed/verbose to have in the README.

Please make a file in `docs/` and add these instructions there.

Author: Will do

Reviewer 3: @whoaitsaimz i'm going to add some details on docs/, let's chat when you're available and see how we can merge our changes

> 1. Go to https://console.firebase.google.com/u/2/.

Reviewer 2: Probably don't add `.` after these lines. Also, drop the `/u/2` from here as that is picking the 2nd logged in google account which isn't going to be universally correct.

```suggestion
1. Go to https://console.firebase.google.com/
```

Author: I pulled all these URLs from original docs and agree they don’t make much sense

2. Click _Add project_.
  * Fill out the name field with something like `example-project`
  * Click _Continue_

> 

Reviewer 2: ```suggestion
```

3. Optional: enable Google Analytics.

  * Click _Continue_
  * Configure Google Analytics
  * Click _Create project_
  * Wait for project to be created
  * Click _Continue_

This part of the setup is complete!

> 

Reviewer 2: ```suggestion
```

### Step 2: Configure Firebase Project Settings

1. Go to https://console.firebase.google.com/ and find the `Example Project` you chose in step 1.

2. From _Project Overview_, hover over the gear icon and click _Project Settings_.

 * In the 'General' tab:

    * Set the public-facing name.

>    * In the ['Users & Permissions'](https://console.firebase.google.com/u/2/project/`example-project`/settings/iam) tab:

Reviewer 2: Don't create links to the admin console like this, they won't work, and needlessly complicates things. Just describe what to do.

Author: It was part of original docs that I copied from, I just moved hyperlink to a new position, but didn’t notice the example-project placeholder till after. Agree we don’t need any of these links really, just descriptions

    * Add any other users who will need "Owner" level access to this project (e.g. a backup owner who can change things if needed, deployments).

>     * Add the users who will need "Editor" level access to this project.

Reviewer 2: These bullet points are all over the place. Needs fixing.

Author: Each main bullet is for the topmost tab. Looked good in my markdown preview but will review again

> 3. Optionally, upgrade your account. Click the gear icon next to _Project Overview_ again, and then select ['Usage and billing'](https://console.firebase.google.com/u/2/project/`example-project`/usage), then select the Blaze plan.

Reviewer 2: Don't create links to the admin console like this, they won't work, and needlessly complicates things. Just describe what to do.

### Step 3: Add a New Web App to the Firebase Project

> 1. Click on _Project Overview_, then again on [_Project Settings_](https://console.firebase.google.com/u/2/project/`example-project`/settings/general).

Reviewer: ```suggestion
1. Click on _Project Overview_, then again on _Project Settings_.
```

> 1. Click on _Project Overview_, then again on [_Project Settings_](https://console.firebase.google.com/u/2/project/`example-project`/settings/general).

Reviewer 2: Don't create links to the admin console like this, they won't work, and needlessly complicates things. Just describe what to do.

>  * In the 'General' tab:

Reviewer: ```suggestion
 * In the _General_ tab:
```

>     * Go to the 'Your Apps' section at the bottom.

Reviewer: ```suggestion
    * Go to the _Your Apps_ section at the bottom.
```

    * Click on the `</>` button to create a new web app
    * Add firebase to your web app
    * Choose an app nickname (eg. `example-project`)
    * Tick 'Also set up Firebase Hosting for this app', leaving the dropdown that appears on the default setting
    * Click _Register app_
    * Wait for the _loading_ spinner to finish
    * Click _Next_ to skip past this screen

>     * Click _Next_ to skip past this screen as well

Reviewer: should we specify here what exact views to skip?

>     * Finally, you can click _Continue to the console_ to skip past this last screen; internal deployments are handled automatically through our CircleCI workflow, so we don't need to deploy manually

Reviewer: should we mention that non-Sparkle guys don't need to worry if they don't use Circle CI unless they want to deploy something? or how to configure everything manually? or whatever else worth mentioning here?

This part of the setup is now complete!

> 

Reviewer 2: Remove extra line
```suggestion
```

### Step 4: Set up Firebase Hosting

> 1. From the Firebase console, within the appropriate project, click on ['Hosting'](https://console.firebase.google.com/u/2/project/`example-project`/hosting) on the left hand menu.

Reviewer: ```suggestion
1. From the Firebase console, within the appropriate project, click on _Hosting_ on the left hand menu.
```

2. Click on _Get started_.

>  * You can skip through these next screens by clicking _Next_
>   * Click _Continue to the console_

Reviewer: ```suggestion
  * You can skip through these next screens by clicking _Next_
  * Click _Continue to the console_
```

> 2. OPTIONAL Add a custom domain.

Reviewer: I believe we should keep same style for "optional" marks. Suggesting the same way as you did it earlier:
```suggestion
2. Optional: Add a custom domain.
```

 * In the hosting dashboard, click _Add custom domain_
 * Enter the details for the domain to use (eg. `example.sparkle.space`)
    **Note:** you will also need to configure the DNS to make this work properly (see below), and this may take 24-48hrs to take effect
 * Don't check 'Redirect `example.sparkle.space` to an existing website'
 * Click _Continue_
 * If you haven't verified your domain before, you'll need to do this here. You will likely need to create a DNS 'TXT' record to prove that you own the domain, containing a `google-site-verification` value
 * You can follow similar steps to the ones described below to do this: _Enter domain --> Verify ownership --> Go live_
 * You will need to enter the provided DNS 'A' records into your domain registrar (eg. GoDaddy) to point your domain at the Firebase servers (record type, host, value)

You can close this page and return to the Firebase console.

> 
> </details>
> 

Reviewer: can't find where we open this tag.
```suggestion
```

Author: I took out DNS settings section. We may want to include in a sub-doc later but it seemed unnecessary for just getting started. Was going to include the GoDaddy example as a dropdown hence the <details> html but that is a relic, good catch!

> 

Reviewer 2: You have a `</details>` here. Why? Where is the start tag? Etc.

### Step 5: Set up Firebase Authentication

> 1. From the Firebase console, within the appropriate project, click on ['Authentication'](https://console.firebase.google.com/u/2/project/`example-project`/authentication) on the left hand menu.

Reviewer: ```suggestion
1. From the Firebase console, within the appropriate project, click on _Authentication_ on the left hand menu.
```

> 1. From the Firebase console, within the appropriate project, click on ['Authentication'](https://console.firebase.google.com/u/2/project/`example-project`/authentication) on the left hand menu.

Reviewer 2: It's probably not worth making this a link since it's not going to work. The way you've combined both the link and the backticks here is wrong too. One or the other, definitely not both.

Author: You’re right and as Viktoriya mentioned the u/2/ part is weird so I’m fine with just a vanilla link to dashboard at the beginning and then text descriptions of where to click

2. Click on _Get Started_.

  * In the ‘Sign-in method’ tab:

>     * Hover over ‘Email/Password’, then click the edit icon (pencil)

Reviewer: ```suggestion
    * Hover over _Email/Password_, then click the edit icon (_pencil_)
```

>     * Select ‘enabled’ (but leave email link ‘passwordless sign in’ disabled)

Reviewer: ```suggestion
    * Toggle on _Enabled_ (but leave email link _passwordless sign in_ disabled)
```

    * Click _Save_
    * Scroll down to 'Authorised domains'
    * Click _Add domain_
    * In the 'Domain' field, enter the custom domain we created above (eg. `example.sparkle.space`)
    * Click _Add_
    * Scroll down to 'Advanced'

>   3. _Note that there is a 'Manage sign-up quota' setting, which defaults to `100` signups per hour from the same IP address. Under most normal cases, this probably doesn't need to be changed, but sometimes (eg. employees all accessing from the same office, therefore having the same IP address), you may need to temporarily increase this limit._

Reviewer: should we include that at all? maybe worth marking that as "optional" ? I've never increased that quota.

Author: Oh good to know. Maybe we can take that whole section and put it in an Advanced Config sub-doc. Thinking Firebase setup may also want to be a dedicated sub doc also

    * To do this, click _Change_
        * Configure 'Sign-ups per hour', 'Start Date', 'Time', and 'Duration (days)' as required
         * 'Sign-ups per hour' must be between `1` and `1000`
         * 'Start Date' must be in the future
         * 'Duration (days)' must be between `1` and `7`
    * Click _Save_
    * Notice the 'Sign-up quota change scheduled' notification

This part of the setup is complete!

### Step 6: Additional Notes

> If you're interested in setting up the project to run in your own environment but you're not sure how to go about it, feel free to [open an issue](https://github.com/sparkletown/sparkle/issues/new) asking for assistance.

Reviewer 2: This line basically existed as a catchall before we had public documentation for this. Not sure if it's going to be relevant anymore once we do.

cd sparkle
```

### Step 2: Install Packages

Install the platform dependencies with `npm`:

> **Important Note:** `npm` v7+ is not supported, it will cause issues with our `package-lock.json`, and you may end up with the wrong dependency versions. Check your current npm version with `npm -v`.

Reviewer 2: "Important" or "Note". Choose one. Both is overkill.

Reviewer 3: fix 6c6c049

```bash
npm install
```

### Step 3: Connect with Firebase

> In the main `sparkle` folder, copy `.env.example` to a new file `.env.local`, and fill it with the appropriate details, most of which you can find in the Firebase dashboard.

Reviewer: should we help everyone here saying where the exact needed evn values can be found?

> In the main `sparkle` folder, copy `.env.example` to a new file `.env.local`, and fill it with the appropriate details, most of which you can find in the Firebase dashboard.

Reviewer 2: ```suggestion
In the main `sparkle` folder, copy `.env.example` to a new file `.env.local`, and fill it with the appropriate details, most of which you can find in the Firebase admin console dashboard.
```

Reviewer 3: fix 6c6c049

You can read more about the various `.env` files that you can use at: https://create-react-app.dev/docs/adding-custom-environment-variables#adding-development-environment-variables-in-env

### Step 4: Start Your Engines

Now you're ready to start the server! ✨

```bash
npm start
```

Once the server is started, your web browser will be opened at http://localhost:3000 (and then it'll be immediately redirected to https://sparklespaces.com/).

To start using the app, you must have a venue setup. Venue URLs take the form of http://localhost:3000/v/{venueName} - replacing `{venueName}` with the existing venue you'd like to use.

> If this is your first time running Sparkle and you are using a fresh Firebase, we have simplified the initial venue setup and admin access using scripts found in the **/scripts/** directory.

Reviewer: should we mention here exact steps to follow? (create a bootstrap venue, add themselves admins, etc)
or at least, tell where these instructions can be found.

While you generally won't need to do this while developing locally, you can manually build the platform assets as follows:

```bash
npm run build
```

> **Note**: You might need to emulate the firebase functions locally before the server can properly start. If you have issues using/editing the actual staging functions, try that.

Reviewer 2: ```suggestion
**Note**: You might need to emulate the firebase functions locally before the server can properly start. If you have issues using/editing your actual deployed environment functions, try that.
```

Reviewer: fixed here: https://github.com/sparkletown/sparkle/pull/1611/commits/c129ebfd44e8b6bb265ae499b7e81964f93b0fcc

> ### Step 5: Configure Firebase Functions

Author: Got it. I’d love to be able to highlight that difference, will revisit

> ### Step 5: Configure Firebase Functions

Reviewer 2: Off the top of my head we want a thorough ‘setting up a new firebase environment’ document (standalone)

then we want a ‘getting started’ that tells you the basic steps of getting your local machine/etc set up, and assumes you already have firebase setup (linking to the other doc for them to follow if they don’t)

> ### Step 5: Configure Firebase Functions

Author: I’m fine with linking to a sub-doc for Firebase setup. The fact that it wasn’t included at all before -and did not have any reference on where to go to find that info (for external users) - is what we need to fix. 

Further delineation of onetime “getting your local machine setup” vs “creating new environment” (with existing Firebase and existing npm and other requirements) sounds great

> Before you run the following steps, you will need to ensure you have access to the Firebase project you want to use. This access can be set up through the Firebase web UI.

Reviewer 2: ```suggestion
Before you run the following steps, you will need to ensure you have access to the Firebase project you want to use. This access can be set up through the Firebase admin console.
```

Reviewer 3: fix 6c6c049

# Login to firebase with the account that has access to this project
npm run firebase login

> # If you are contributing to the Sparkle main code base, switch to the 'staging' project, otherwise switch to 'example-project' or whichever environment you are developing against

Reviewer 3: fix 6c6c049

tmp/firestore_export
```

---

> <!-- section 3 -->

Reviewer 2: Why adding these?

## Part Three: Configure Third-Party Integrations

### Optional: Twilio

TODO: Add steps.

> ### Optional: Stripe

Author: Ok... got it and stripe was causing some issues anyway. Worth trying to clean up main code base and rid it of Stripe completely? Since we don’t really have a working implementation yet... to my knowledge

> ### Optional: Stripe

Reviewer 2: Not in time for this event. It’s a known tech debt thing, nothing needs to be done for setup, it will make some slight noise in the console. Far higher priority things to allocate time to, zero risk for not cleaning it up now.

const STRIPE_ENDPOINT_KEY = `${YOUR_LOCAL_SIGNING_SECRET_KEY}`;
```

---

> <!-- section 4 -->

Reviewer 2: Why adding these?

Author: They were already in there breaking up Part 1. I assumed for some kind of markdown section breaks or something but if they have no purpose we can just remove altogether

Reviewer 2: The —- will be a horizontal rule. Don’t think the comments do anything but

- Pushing the `master` branch to any of our other configured production branches will deploy to that environment

## Obtaining email addresses from firebase

WARNING: Only email people with consent. Permission based marketing is the best way to grow your email list.

```
$ firebase auth:export --project co-reality-map auth.json
$ jq '.users[] | "\(.createdAt) \(.email)"' auth.json|awk '$2!="null\"" {print $0}'|sed -e 's/^"//' -e 's/"$//'|awk '{print $1"\t"$2}'|pbcopy
```

Paste into a google sheet. Leftmost column is the [UNIX timestamp](https://en.wikipedia.org/wiki/Unix_time) of the account creation. This can be converted easily:

1. Create a new column to the right of the timestamp
2. Use the formula: `=A4/86400000 + DATE(1970,1,1)`
3. Fill down to get all values
4. Sort by timestamp or the calculated, human date
5. Consider not adding any with a "+"
6. Be mindful this is PII (personally identifiable information) so handle it carefully and treat it as sensitive. It may be subject to the GDPR data privacy requirements in the EU and the CCPA privacy laws in California.
7. Share the Google sheet, ready to import the users into our other email lists.

> 8. Email the new folks, welcome them, and say thanks for coming to the party!

Author: It was but I put in internal doc

> 8. Email the new folks, welcome them, and say thanks for coming to the party!

Reviewer 2: Could be useful knowledge for others publicly that want to use the open source project. My ultimate goal is to have no internal docs and for everything to be fully public. But for now, internal is a stopgap to remove the need to consider so carefully about privacy/etc things in documentation. Given this already existed publicly, presumably it should stay as such.

A new file in docs/ would be good; within the main repo. Doesn’t need to be linked from the readme at all

> 8. Email the new folks, welcome them, and say thanks for coming to the party!

Reviewer 3: fix 6c6c049
